README.md

# matrixio-creator-init

## matrix-creator-start-services
![Build Status](https://drone.matrix.one/api/badges/matrix-io/matrix-creator-init/status.svg)

Source for the matrix-creator-init Debian package. This package programs the FPGA and the SAM3 MCU when the system starts.

## Installation
```
# Add repo and key
curl https://apt.matrix.one/doc/apt-key.gpg | sudo apt-key add -
echo "deb https://apt.matrix.one/raspbian $(lsb_release -sc) main" | sudo tee /etc/apt/sources.list.d/matrixlabs.list

# Update packages and install
sudo apt-get update
sudo apt-get upgrade

# Installation
sudo apt install matrixio-creator-init
```
## Documentation
https://matrix-io.github.io/matrix-documentation


debian/changelog
matrixio-creator-init (0.4.17) unstable; urgency=medium

  * Support for Raspberry Pi 4 (sysfs workaround)

 --  Andres Calderon <andres.calderon@admobilize.com>  Tue, 12 Ago 2019 18:00:00 +0000

matrixio-creator-init (0.4.16) unstable; urgency=medium

  * MCU blob updated (humidity fixed)

 --  Andres Calderon <andres.calderon@admobilize.com>  Tue, 9 Jul 2019 16:00:00 +0000

matrixio-creator-init (0.4.15) unstable; urgency=medium

  * Fix script to add new setings in /boot/config.txt

 --  Leonardo Urrego <leonardo.urrego@admobilize.com>  Thu, 27 Jun 2019 15:00:00 +0000
 
 matrixio-creator-init (0.4.14) unstable; urgency=medium

  * Buster support added (Jessie deprecated). 

 --  Andres Calderon <andres.calderon@matrixlabs.ai>  Tue, 25 Jun 2019 15:00:00 +0000

matrixio-creator-init (0.4.13) unstable; urgency=medium

  * Fix in Playback core module. It is now compatible with signed 16bit PCM  

 --  Yoel Ramos Castillo <yoel.castillo@matrixlabs.ai>  Tue, 13 Nov 2018 15:00:00 +0000

matrixio-creator-init (0.4.12) unstable; urgency=medium

  * GPIO Core MATRIX Voice

 --  Kevin Patino <kevin.patino@admobilize.com>  Thu, 5 Jun 2018 15:00:00 +0000

matrixio-creator-init (0.4.11) unstable; urgency=medium

  * Bug Fix: Remove pop in playback core.

 --  Kevin Patino <kevin.patino@admobilize.com>  Fri, 1 Jun 2018 15:00:00 +0000

matrixio-creator-init (0.4.10) unstable; urgency=medium

  * Support Zwave GPIO Manager.

 --  Kevin Patino <kevin.patino@admobilize.com>  Fri, 25 May 2018 15:00:00 +0000

matrixio-creator-init (0.4.9) unstable; urgency=medium

  * Support matrixio-microphone-core: Update FPGA firmware.

 --  Kevin Patino <kevin.patino@admobilize.com>  Fri, 18 May 2018 15:00:00 +0000

matrixio-creator-init (0.4.8) unstable; urgency=medium

  * Support libmatrixio-creator-hal unification release.

 --  Kevin Patino <kevin.patino@admobilize.com>  Mon, 9 Apr 2018 15:00:00 +0000

matrixio-creator-init (0.4.7) unstable; urgency=medium

  * Fix libmatrixio-creator-hal as a dependency.

 --  Kevin Patino <kevin.patino@admobilize.com>  Mon, 2 Apr 2018 15:00:00 +0000

matrixio-creator-init (0.4.6) unstable; urgency=medium

  * Support 96KHz in MATRIX Devices.
  * Add file to configure spidev buffer in 12000.
  * Remove asound.conf.
  * Make compatible with libmatrixio-creator-hal v0.3.0

 --  Kevin Patino <kevin.patino@admobilize.com>  Mon, 26 Feb 2018 16:45:00 +0000

matrixio-creator-init (0.4.5) unstable; urgency=medium

  * Bug Fix: Em358 check process has been fixed

 --  Kevin Patino <kevin.patino@admobilize.com>  Wed, 21 Feb 2018 16:45:00 +0000

matrixio-creator-init (0.4.4) unstable; urgency=medium

  * MATRIX Creator hot fix

 --  Kevin Patino <kevin.patino@admobilize.com>  Wed, 14 Feb 2018 11:45:00 +0000

matrixio-creator-init (0.4.3) unstable; urgency=medium

  * Bug Fix: Set a volume gain the Micrphone Array FPGA
  * Update MATRIX Creator FPGA in order to work with the recently updates.

 --  Kevin Patino <kevin.patino@admobilize.com>  Wed, 14 Feb 2018 11:45:00 +0000

matrixio-creator-init (0.4.2) unstable; urgency=medium

  * Support Audio Output in MATRIX Voice
  * Adding config file for ESP32_RESET
  * Bug Fix: MATRIX Creator GPIO controller

 --  Kevin Patino <kevin.patino@admobilize.com>  Mon, 12 Feb 2018 10:32:00 +0000

matrixio-creator-init (0.4.1) unstable; urgency=medium

  * Support write/read functions at the same time ESP32 and Raspberry - MATRIX Voice
  * Use the 0.2.0 Version of libmatrixio-creator-hal - memory maps has been changed
  * Update FPGA firmware - MATRIX Voice - MATRIX Creator

 --  Kevin Patino <kevin.patino@admobilize.com>  Tue, 30 Jan 2018 11:24:00 +0000


matrixio-creator-init (0.4.0) unstable; urgency=medium

  * Adding MATRIX Voice ESP32 tool scripts
  * Adding MATRIX Voice FPGA Support for ESP32

 --  Kevin Patino <kevin.patino@admobilize.com>  Fri, 22 Nov 2017 17:24:00 +0000


matrixio-creator-init (0.3.5) unstable; urgency=medium

  * Adding MATRIX Voice init configuration
  * Adding MATRIX Creator indentify number in FPGA software

 --  Kevin Patino <kevin.patino@admobilize.com>  Fri, 22 Nov 2017 17:24:00 +0000


matrixio-creator-init (0.3.4) unstable; urgency=medium

  * Adding WDT in MCU firmware
  * Improve UART implementation in FPGA core

 --  Kevin Patino <kevin.patino@admobilize.com>  Fri, 22 Nov 2017 17:24:00 +0000

matrixio-creator-init (0.3.3) unstable; urgency=medium

  * Multiple frequency support in micarray has been supported.
  * Support NFC Hardware layer has been added
  * Zwave setup in ttyS0
  * Remove LIRC support 

 --  Kevin Patino <kevin.patino@admobilize.com>  Fri, 15 Sep 2017 19:26:10 +0000

matrixio-creator-init (0.3.2) unstable; urgency=medium

  * LIRC has been removed

 --  Andres Calderon <andres.calderon@admobilize.com>  Thu, 31 Aug 2017 13:11:16 +0000

matrixio-creator-init (0.3.1) unstable; urgency=medium

  * Fixing Unstable behavior in SAM3S

 --  Kevin Patino <kevin.patino@admobilize.com>  Wed, 30 Aug 2017 21:37:36 +0000

matrixio-creator-init (0.3.0) unstable; urgency=medium

  * Read process for EM358 has been updated.

 -- Kevin Patino <kevin.patino@admobilize.com>  Mon, 27 Feb 2017 16:11:53 +0000

matrixio-creator-init (0.2-23) unstable; urgency=medium

  * Adding Zigbee pre-installation scripts 
  * LIRC add custom remotes 

 --  <pi@raspberrypi>  Mon, 23 Jan 2017 07:09:29 -0900

matrixio-creator-init (0.2-22) unstable; urgency=medium

  * Uses  sysfs-gpio to program the SAM3.
  * This makes it work with the last Raspbian image.

 --  <pi@raspberrypi>  Wed, 07 Dec 2016 14:27:37 -0900

matrixio-creator-init (0.2-21) unstable; urgency=medium

  * Fix for https://github.com/matrix-io/matrix-creator-init/issues/10

 --  <pi@raspberrypi>  Fri, 14 Oct 2016 10:51:43 -0800

matrixio-creator-init (0.2-20) unstable; urgency=medium

  * Add raspicam.conf to etc/modules-load.d.

 --  <pi@raspberrypi>  Fri, 14 Oct 2016 10:51:43 -0800

matrixio-creator-init (0.2-19) unstable; urgency=medium


sam3-program.bash
#!/bin/bash

LOG_FILE=/tmp/sam3-program.log

function super_reset()
{
  # gpio19 - EM358 nBOOTMODE (active low)
  # gpio18 - mcu power
  # gpio20 - EM_NRST - EM3588 RESET (active low)
  # gpio23 - EM3588 POWER ENABLE

  # Set out mode  
  for j in 18 19 20 23
  do
    echo out > /sys/class/gpio/gpio$j/direction
  done

  for k in 4 17 22 27
  do
    echo in > /sys/class/gpio/gpio$k/direction
  done

  #Power EM_358 OFF 
  echo 1 > /sys/class/gpio/gpio18/value
  echo 1 > /sys/class/gpio/gpio19/value
  echo 1 > /sys/class/gpio/gpio20/value
  echo 0 > /sys/class/gpio/gpio23/value
  sleep 0.5

  #Power ON
  echo 1 > /sys/class/gpio/gpio23/value
  sleep 0.5

  echo 0 > /sys/class/gpio/gpio18/value
  echo 0 > /sys/class/gpio/gpio19/value
  echo 0 > /sys/class/gpio/gpio20/value

  sleep 0.5
  echo 1 > /sys/class/gpio/gpio18/value
  echo 1 > /sys/class/gpio/gpio20/value

  sleep 0.5
  echo 1 > /sys/class/gpio/gpio19/value
}

function reset_mcu() {
  if [ ! -d /sys/class/gpio/gpio18 ]
  then
    echo 18 > /sys/class/gpio/export
  fi
  echo out > /sys/class/gpio/gpio18/direction
  echo 1 > /sys/class/gpio/gpio18/value
  echo 0 > /sys/class/gpio/gpio18/value
  echo 1 > /sys/class/gpio/gpio18/value
}

function try_program() {
  reset_mcu
  sleep 0.1

  RES=$(openocd -f cfg/sam3s_rpi_sysfs.cfg 2>&1 | tee ${LOG_FILE} | grep wrote | wc -l)
  echo $RES
  
  sleep 0.5
  reset_mcu  
}

function enable_program() {
  echo 1 > /sys/class/gpio/gpio19/value
  echo 0 > /sys/class/gpio/gpio20/value
  echo 1 > /sys/class/gpio/gpio20/value
  echo "Running the program instead of the bootloader" 
}

function check_firmware() {
 COMPARE_VERSION=$(diff <(./firmware_info | grep MCU) <(cat mcu_firmware.version)|wc -l)

 if [ "$COMPARE_VERSION" == "0" ];then
  echo 1
 else #failed
  echo 0 
 fi
}

for i in 4 17 18 19 20 22 23 27
do
  if [ ! -d /sys/class/gpio/gpio$i ]
  then
    echo $i > /sys/class/gpio/export
  fi
done

super_reset 
reset_mcu

CHECK=$(check_firmware)
if [ "$CHECK" == "1" ]
then
  reset_mcu
  enable_program
  echo "SAM3 MCU was programmed before. Not programming it again."
  exit 0
fi
enable_program
count=0
while [  $count -lt 30 ]; do
  TEST=$(try_program)
  if [ "$TEST" == "1" ];then
        CHECK=$(check_firmware)
        if [ "$CHECK" == "1" ];then
          echo "****  SAM3 MCU programmed!"
          reset_mcu
          exit 0
        fi
   fi
  let count=count+1
done
echo "**** Could not program SAM3 MCU, you must be check the logfile ${LOG_FILE}"
exit 1
